packages/enhanced-img/src/preprocessor.js

```diff
@@ -0,0 +1,188 @@
+import MagicString from 'magic-string';
+import { parse } from 'svelte-parse-markup';
+import { walk } from 'svelte/compiler';
+
+const ASSET_PREFIX = '___ASSET___';
+
+// TODO: expose this in vite-imagetools rather than duplicating it
+const OPTIMIZABLE = /^[^?]+\.(avif|heif|gif|jpeg|jpg|png|tiff|webp)(\?.*)?$/;
+
+/**
+ * @returns {import('svelte/types/compiler/preprocess').PreprocessorGroup}
+ */
+export function image() {
+	return {
+		markup({ content, filename }) {
+			const s = new MagicString(content);
+			const ast = parse(content, { filename });
+
+			// Import path to import name
+			// e.g. ./foo.png => ___ASSET___0
+			/** @type {Map<string, string>} */
+			const imports = new Map();
+
+			/**
+			 * @param {import('svelte/types/compiler/interfaces').TemplateNode} node
+			 * @param {{ type: string, start: number, end: number, raw: string }} src_attribute
+			 */
+			function update_element(node, src_attribute) {
+				if (src_attribute.type === 'MustacheTag') {
+					const src_var_name = content
+						.substring(src_attribute.start + 1, src_attribute.end - 1)
+						.trim();
+					s.update(node.start, node.end, dynamic_img_to_picture(content, node, src_var_name));
+					return;
+				}
+
+				let url = src_attribute.raw.trim();
+
+				const sizes = get_attr_value(node, 'sizes');
+				const width = get_attr_value(node, 'width');
+				url += url.includes('?') ? '&' : '?';
+				if (sizes) {
+					url += 'imgSizes=' + encodeURIComponent(sizes.raw) + '&';
+				}
+				if (width) {
+					url += 'imgWidth=' + encodeURIComponent(width.raw) + '&';
+				}
+				url += 'enhancedImg';
+
+				let import_name = '';
+				if (imports.has(url)) {
+					import_name = /** @type {string} */ (imports.get(url));
+				} else {
+					import_name = ASSET_PREFIX + imports.size;
+					imports.set(url, import_name);
+				}
+
+				if (OPTIMIZABLE.test(url)) {
+					s.update(node.start, node.end, img_to_picture(content, node, import_name));
+				} else {
+					// e.g. <img src="./foo.svg" /> => <img src="{___ASSET___0}" />
+					s.update(src_attribute.start, src_attribute.end, `{${import_name}}`);
+				}
+			}
+
+			let ignore_next_element = false;
+
+			// @ts-ignore
+			walk(ast.html, {
```

Comment:
oh. huh. i had forgotten we exposed this from `svelte/compiler`. we no longer use `estree` inside the compiler, we use its successor https://github.com/Rich-Harris/zimmerframe. i'm not sure if we're likely to expose a backwards compatible `walk` function — probably not — so it might be better to use zimmerframe here